test/FTSAlgorithmsTest.cpp

```diff
@@ -30,209 +30,191 @@ TEST(FTSAlgorithmsTest, filterByRangeTest) {
-  ASSERT_EQ(0u, resCids.size());
-  ASSERT_EQ(0u, resScores.size());
-
-  cids2.push_back(T(2));
-  scores2.push_back(1);
-  FTSAlgorithms::intersectTwoPostingLists(cids1, scores1, cids2, scores2,
-                                          resCids, resScores);
-  ASSERT_EQ(1u, resCids.size());
-  ASSERT_EQ(1u, resScores.size());
-
-  cids2.push_back(T(4));
-  scores2.push_back(1);
-  FTSAlgorithms::intersectTwoPostingLists(cids1, scores1, cids2, scores2,
-                                          resCids, resScores);
-  ASSERT_EQ(1u, resCids.size());
-  ASSERT_EQ(1u, resScores.size());
+  matchingContextsWep.cids_ = {T(0), T(2), T(2)};
+  matchingContextsWep.wids_ = {{1, 4, 8}};
+  matchingContextsWep.scores_ = {1, 1, 1};
+
+  resultWep = FTSAlgorithms::crossIntersect(matchingContextsWep, eBlockWep);
+  ASSERT_EQ(4u, resultWep.wids_[0].size());
+  ASSERT_EQ(4u, resultWep.cids_.size());
+  ASSERT_EQ(4u, resultWep.eids_.size());
+  ASSERT_EQ(4u, resultWep.scores_.size());
+  ASSERT_EQ(8u, resultWep.wids_[0][1]);
 };
 
-TEST(FTSAlgorithmsTest, intersectKWayTest) {
+TEST(FTSAlgorithmsTest, crossIntersectKWayTest) {
   vector<Index::WordEntityPostings> wepVecs;
   Index::WordEntityPostings resultWep;
 
-  vector<Score> fourScores;
-  fourScores.push_back(1);
-  fourScores.push_back(1);
-  fourScores.push_back(1);
-  fourScores.push_back(1);
-
-  vector<TextRecordIndex> cids1;
-  cids1.push_back(T(0));
-  cids1.push_back(T(1));
-  cids1.push_back(T(2));
-  cids1.push_back(T(10));
-
   Index::WordEntityPostings wep1;
-  wep1.cids_ = cids1;
-  wep1.scores_ = fourScores;
-  wepVecs.push_back(wep1);
+  wep1.scores_ = {1, 1, 1, 1};
+  wep1.cids_ = {T(0), T(1), T(2), T(10)};
+  wep1.wids_ = {{3, 2, 5, 3}};
 
-  wep1.cids_[2] = T(8);
-  wepVecs.push_back(wep1);
+  Index::WordEntityPostings wep2;
+  wep2.scores_ = {1, 1, 1, 1};
+  wep2.cids_ = {T(0), T(0), T(0), T(10)};
+  wep2.wids_ = {{8, 7, 6, 9}};
+
+  Index::WordEntityPostings wep3;
+  wep3.scores_ = {1, 1, 1, 3};
+  wep3.cids_ = {T(0), T(6), T(8), T(10)};
+  wep3.wids_ = {{23, 22, 25, 23}};
 
-  wep1.cids_[1] = T(6);
-  wep1.scores_[3] = 3;
   wepVecs.push_back(wep1);
+  wepVecs.push_back(wep2);
+  wepVecs.push_back(wep3);
 
   // No eids / no special case
-  resultWep = FTSAlgorithms::intersectKWay(wepVecs, nullptr);
-  ASSERT_EQ(2u, resultWep.cids_.size());
+  resultWep = FTSAlgorithms::crossIntersectKWay(wepVecs, nullptr);
+  ASSERT_EQ(4u, resultWep.cids_.size());
   ASSERT_EQ(0u, resultWep.eids_.size());
-  ASSERT_EQ(2u, resultWep.scores_.size());
+  ASSERT_EQ(4u, resultWep.scores_.size());
+  ASSERT_EQ(3u, resultWep.wids_.size());
+  ASSERT_EQ(4u, resultWep.wids_[0].size());
   ASSERT_EQ(3u, resultWep.scores_[0]);
-  ASSERT_EQ(5u, resultWep.scores_[1]);
+  ASSERT_EQ(3u, resultWep.scores_[1]);
+  ASSERT_EQ(5u, resultWep.scores_[3]);
+  ASSERT_EQ(23u, resultWep.wids_[2][1]);
+  ASSERT_EQ(7u, resultWep.wids_[1][1]);
+  ASSERT_EQ(T(0), resultWep.cids_[2]);
```

Comment:
This can be much cleaner using `gmock` matchers like `::testing::ElementsAreArray` etc
(look for Gmock matchers, or look in some of our other tests).